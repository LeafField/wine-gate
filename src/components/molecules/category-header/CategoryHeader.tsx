"use client";
import React, {
  ChangeEvent,
  Dispatch,
  FC,
  SetStateAction,
  useState,
} from "react";
import { NativeSelect } from "@mantine/core";

type Props = {
  title: string;
  sort?: string;
  setSort?: Dispatch<SetStateAction<string>>;
};

const CategoryHeader: FC<Props> = ({ title, sort, setSort }) => {
  const changeHandler = (event: ChangeEvent<HTMLSelectElement>) => {
    if (!setSort) return;
    setSort(event.target.value);
  };

  return (
    <div className="mx-4 flex flex-col-reverse justify-between gap-y-2 border-b-2 border-red sort tablet:mx-0 tablet:w-[calc(var(--vw)-401px)] tablet:flex-row tablet:items-center">
      <h2 className="text-heading2">{title}</h2>
      <NativeSelect
        data={[
          { value: "popular", label: "人気順" },
          { value: "new", label: "新着順" },
          { value: "inexpensive", label: "価格が安い順" },
        ]}
        className="block w-48"
        value={sort}
        onChange={changeHandler}
        id="sort-select"
      />
    </div>
  );
};

export default CategoryHeader;
