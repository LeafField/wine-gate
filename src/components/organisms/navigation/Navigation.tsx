"use client";
import React from "react";
import CategoryPanel from "../../molecules/category-panel/CategoryPanel";
import SearchPanel from "../../molecules/search-panel/SearchForm";
import UserPanel from "../../molecules/user-panel/UserPanel";
import { selectDummyData } from "../../../utils/dummyData";
import { categoryDummyData } from "../../../utils/dummyData";
import NavMenu from "../../molecules/nav-menu/NavMenu";
import { useStore } from "../../../store";

const Navigation = () => {
  const { menu } = useStore();
  return (
    <nav className="nav">
      <div
        className={`${menu === "close" ? "invisible" : "show-panel"} z-nav tablet:visible tablet:static tablet:space-y-6`}
        id="navigation"
      >
        <UserPanel />
        <CategoryPanel categories={categoryDummyData} />
        <SearchPanel selectData={selectDummyData} />
      </div>
      <NavMenu />
    </nav>
  );
};

export default Navigation;
