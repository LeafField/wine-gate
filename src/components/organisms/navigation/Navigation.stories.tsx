import { Meta, StoryObj } from "@storybook/react";
import Navigation from "./Navigation";
import { spyOn, within } from "@storybook/test";
import { supabase } from "../../../utils/supabase";

const meta = {
  title: "organisms/Navigation",
  component: Navigation,
  parameters: {
    nextjs: {
      appDirectory: true,
    },
  },
} as Meta<typeof Navigation>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  play: async ({ canvasElement }) => {
    within(canvasElement);
    spyOn(supabase.auth, "signOut").mockImplementationOnce(() => {
      console.log("signOut");
      return Promise.resolve({ error: null, data: null });
    });
  },
};
