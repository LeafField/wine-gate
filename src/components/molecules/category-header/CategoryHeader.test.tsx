import { screen, render, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import CategoryHeader from "./CategoryHeader";
import * as nextRouter from "next/navigation";
import { MantineProvider } from "@mantine/core";

jest.mock("next/navigation");

describe("CategoryHeaderの単体テスト", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  afterAll(() => {
    jest.restoreAllMocks();
  });

  test("CategoryHeaderが正しく表示される", () => {
    jest
      .spyOn(nextRouter, "useParams")
      .mockReturnValue({ category: ["sober", "new", "1"] });
    render(<CategoryHeader title="新着ワイン" />, {
      wrapper: MantineProvider,
    });
    expect(screen.getByText("新着ワイン")).toBeInTheDocument();
    expect(screen.getByDisplayValue("新着順")).toBeInTheDocument();
  });

  test("CategoryHeaderのセレクトボックスの選択肢が正しく表示される", async () => {
    jest
      .spyOn(nextRouter, "useParams")
      .mockReturnValue({ category: ["sober", "new", "1"] });
    render(<CategoryHeader title="新着ワイン" />, {
      wrapper: MantineProvider,
    });

    await userEvent.click(screen.getByDisplayValue("新着順"));
    await waitFor(() => {
      expect(screen.getByText("新着順")).toBeInTheDocument();
      expect(screen.getByText("価格が安い順")).toBeInTheDocument();
      expect(screen.getByText("人気順")).toBeInTheDocument();
    });
  });

  test("CategoryHeaderのセレクトボックスの選択肢を変更すると、router.pushが呼ばれる", async () => {
    jest
      .spyOn(nextRouter, "useParams")
      .mockReturnValue({ category: ["sober", "new", "1"] });
    const pushMock = jest.fn();
    jest.spyOn(nextRouter, "useRouter").mockReturnValue({
      push: pushMock,
    } as any);
    render(<CategoryHeader title="新着ワイン" />, {
      wrapper: MantineProvider,
    });

    await userEvent.selectOptions(
      screen.getByDisplayValue("新着順"),
      "popular",
    );
    expect(pushMock).toHaveBeenCalledWith("/category/sober/popular/1");
  });

  test("params.category[1]がundefinedの場合、デフォルトで新着順が選択される", () => {
    jest
      .spyOn(nextRouter, "useParams")
      .mockReturnValue({ category: ["sober"] });
    render(<CategoryHeader title="新着ワイン" />, {
      wrapper: MantineProvider,
    });
    expect(screen.getByDisplayValue("新着順")).toBeInTheDocument();
  });

  test("スナップショットテスト", () => {
    const { container } = render(<CategoryHeader title="新着ワイン" />, {
      wrapper: MantineProvider,
    });
    expect(container).toMatchSnapshot();
  });
});
