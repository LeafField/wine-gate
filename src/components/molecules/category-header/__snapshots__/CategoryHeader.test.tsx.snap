// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryHeaderの単体テスト スナップショットテスト 1`] = `
<div>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
  </style>
  <div
    class="flex flex-col-reverse justify-between gap-y-2 border-b-2 border-red tablet:flex-row tablet:items-center"
  >
    <h2
      class="text-heading2"
    >
      新着ワイン
    </h2>
    <div
      class="block w-48 m-46b77525 mantine-InputWrapper-root mantine-NativeSelect-root"
    >
      <div
        class="m-6c018570 mantine-Input-wrapper mantine-NativeSelect-wrapper"
        data-pointer="true"
        data-variant="default"
        data-with-right-section="true"
        style="--input-right-section-pointer-events: none;"
      >
        <select
          aria-invalid="false"
          class="m-8fb7ebe7 mantine-Input-input mantine-NativeSelect-input"
          data-variant="default"
          id="sort-select"
        >
          <option
            value="new"
          >
            新着順
          </option>
          <option
            value="inexpensive"
          >
            価格が安い順
          </option>
          <option
            value="popular"
          >
            人気順
          </option>
        </select>
        <div
          class="m-82577fc2 mantine-Input-section mantine-NativeSelect-section"
          data-position="right"
        >
          <svg
            class="m-2943220b mantine-ComboboxChevron-chevron"
            data-combobox-chevron="true"
            fill="none"
            viewBox="0 0 15 15"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M4.93179 5.43179C4.75605 5.60753 4.75605 5.89245 4.93179 6.06819C5.10753 6.24392 5.39245 6.24392 5.56819 6.06819L7.49999 4.13638L9.43179 6.06819C9.60753 6.24392 9.89245 6.24392 10.0682 6.06819C10.2439 5.89245 10.2439 5.60753 10.0682 5.43179L7.81819 3.18179C7.73379 3.0974 7.61933 3.04999 7.49999 3.04999C7.38064 3.04999 7.26618 3.0974 7.18179 3.18179L4.93179 5.43179ZM10.0682 9.56819C10.2439 9.39245 10.2439 9.10753 10.0682 8.93179C9.89245 8.75606 9.60753 8.75606 9.43179 8.93179L7.49999 10.8636L5.56819 8.93179C5.39245 8.75606 5.10753 8.75606 4.93179 8.93179C4.75605 9.10753 4.75605 9.39245 4.93179 9.56819L7.18179 11.8182C7.35753 11.9939 7.64245 11.9939 7.81819 11.8182L10.0682 9.56819Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
